### PR TITLE
feat(temurin) introduces a new source of type "Temurin"

### LINKED ISSUE #1733: Feature Request: Support for version pattern in autodiscovery for docker image eclipse-temurin
### Is your feature request related to a problem?

The eclispe temurin version  pattern (eg eclipse-temurin:17.0.6_10-jdk) is not understood by updatecli's version parser

### Solution you'd like

Support for patterns like eclipse-temurin:17.0.6_10-jdk

### Alternatives you've considered

Implemented manually writing source and target rules.
This bloats our ci pipeline enormously, though.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dduportal**
Fix #1733

This PR introduces a new resource of kind `temurin`. Only source is supported.

It allows retrieving [Adoptium ("Temurin JDK")](https://adoptium.net/) JDK version, Installer URL, Checksum URL or Signature URL based on different filtering criterias.
Implementation uses the [Adoptium API](https://api.adoptium.net/q/swagger-ui/#/Assets/searchReleases) under the hood.

Quick examples (Documentation on the website is needed):

```
sources:
  ## Returns last LTS version
  getLastDefaultVersion:
    kind: temurin

  ## Returns last feature release (explicit)
  getLastFeatureVersion:
    kind: temurin
    spec:
      releaseline: feature # Can be 'lts' (default), 'feature'

  ## Returns last LTS Nightly
  getLastLTSNightlyVersion:
    kind: temurin
    spec:
      releaseline: lts
      releasetype: ea # Can be 'ga' (default), 'ea'

  # Returns the installer URL for the latest x64 Linux JDK in the LTS version line
  getLatestLTSLinuxAMD64InstallerURL:
    kind: temurin
    spec:
      result: installer_url # Can be 'version' (default), 'installer_url', 'checksum_url', 'signature_url'

  ## Returns the installer checksum URL for the latest x64 Linux JDK in the LTS version line
  getLatestLTSLinuxAMD64ChecksumUrl:
    kind: temurin
    spec:
      result: checksum_url # Can be 'version' (default) 'installer_url', 'checksum_url', 'signature_url'

  ## Returns the installer download URL for the latest x64 Windows JDK in the LTS version line
  getInstallerUrlJDK17WindowsAMD64:
    kind: temurin
    spec:
      result: installer_url
      featureversion: 17 # Major Java release (integer) to filter on.
      operatingsystem: windows # Follows Temurin conventions: can be 'linux' (default), 'alpine-linux', 'windows', 'mac', 'solaris', 'aix"

  ## Returns Checksum URL for latest JRE on the s390x Linux 17 line
  getChecksumJRE17Linuxs390x:
    kind: temurin
    spec:
      featureversion: 17
      operatingsystem: linux
      architecture: s390x # Follows Temurin conventions: can be 'x64' (default), 'x86', 'x32', 'ppc64', 'ppc64le', 's390x', 'aarch64', 'arm', 'sparcv9', 'riscv64'
      imagetype: jre  # Follows Temurin conventions: can be 'jdk' (default), 'jre', 'testimage', 'debugimage', 'staticlibs', 'sources', 'sbom'
      result: checksum_url

  ## Returns Signature URL of the latest JRE for the Windows 17.0.9+9 release
  getSignatureUrlCustomVersionWindows:
    kind: temurin
    spec:
      featureversion: 17
      operatingsystem: windows
      result: signature_url
```

## Test

To test this pull request:

- Unit tests with a mocked HTTP client, covering ~85% lines. Of course it's not real life test but it helps covering a lot of code conditional branches and it won't break with Adoptium new releases:

```shell
cd ./pkg/plugins/resources/temurin
go test ./... -short
```

- End to end test, serving as a "real life test". Note it has conditions to test that any kind of returned URL exists so it survives new releases while being a real life case:

```shell
go build -o ./dist/updatecli
./dist/updatecli diff --config ./e2e/updatecli.d/success.d/temurin/sources.yaml
```

### CODE REVIEW

**Comment by gounthar** [comment]
Excellent news, thanks a bunch! 🥳

File: pkg/plugins/resources/temurin/spec.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+type Spec struct {
+	// ReleaseLine specifies the type of Temurin Release to retrieve: 'lts' (default), feature.
+	ReleaseLine string `yaml:",omitempty"`
+	// ReleaseType specifies the type of Temurin release: 'ga' (default), 'ea' (nightly releases).
+	ReleaseType string `yaml:",omitempty"`
+	// FeatureVersion specifies the Java Feature version (major) to filter results.
+	FeatureVersion int `yaml:",omitempty"`
+	// Result specifies the type of result returned by the resource: 'version' (default), installer_url, checksum/
+	Result string `yaml:",omitempty"`
```
Comment (olblak) [inline pkg/plugins/resources/temurin/spec.go:11]: I like that you allow different kind of result.
It's something that I should have for other plugins like gittag should be able to return a git tag or hash associated to that tag

File: pkg/plugins/resources/temurin/spec.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package temurin
+
+type Spec struct {
+	// ReleaseLine specifies the type of Temurin Release to retrieve: 'lts' (default), feature.
```
Comment (olblak) [inline pkg/plugins/resources/temurin/spec.go:4]: What would you think to split the documentation on mutliple line
```suggestion
	// ReleaseLine specifies the type of Temurin Release to retrieve.
	//
	// default: "lts"
	// 
	// Allowed values:
	// * "lts"
	// * "feature"
```

File: pkg/plugins/resources/temurin/api.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+		return result, fmt.Errorf("[temurin] No release found matching provided criteria. Use '--debug' to get details.")
+	}
+
+	// Return only the most recent, e.g. the first one (sort is DESC in the URL)
+	return apiResult.Releases[0], nil
+}
+
+func (t Temurin) apiGetInstallerUrl(releaseName string) (result string, err error) {
+	apiEndpoint := fmt.Sprintf(
+		"%s/%s/%s/%s/%s/hotspot/normal/eclipse?project=jdk",
```
Comment (olblak) [inline pkg/plugins/resources/temurin/api.go:224]: Stupid question, is it normal that "jdk" is hardcoded?
Would it make sense to use jre or sbom?

Comment (olblak) [reply pkg/plugins/resources/temurin/api.go:224]: Differently asked, could we configure different project?

Comment (dduportal) [reply pkg/plugins/resources/temurin/api.go:224]: Good question. I went for `jdk` as it's the only project which returned HTT/200 answers in my tests.

But there are other values available: worth adding them in the spec

From https://api.adoptium.net/q/swagger-ui/#/Binary/getBinaryByVersion: 

Capture d’écran 2024-08-26 à 12 38 29

File: pkg/plugins/resources/temurin/spec.go (hunk 2 above)
Comment (dduportal) [reply pkg/plugins/resources/temurin/spec.go:4]: Can you help me / remind me how is this multiline syntax works?

I'll update the other spec attributes on the same model in any case

File: pkg/plugins/resources/temurin/api.go (hunk 3 above)
Comment (dduportal) [reply pkg/plugins/resources/temurin/api.go]: Updated the PR with the support for "Project" attribute

**Comment by dduportal** [comment]
PR updates with:
- Support for the "Project" attribute
- Multi line comment for Spec attributes, following @olblak suggestion
- Simplified the condition of the e2e test (not running faster, but easier to read and maintain)

File: pkg/plugins/resources/temurin/spec.go [hunk 4]
```diff
@@ -1,26 +1,94 @@
...
+	// ReleaseType specifies the type of Temurin release to retrieve.
+	//
+	// default: "ga"
+	//
+	// Allowed values:
+	// * "ga" (General Availability)
+	// * "ea" (Early Availability, e.g. nightly builds)
 	ReleaseType string `yaml:",omitempty"`
-	// FeatureVersion specifies the Java Feature version (major) to filter results.
+	// FeatureVersion specifies the Major Java version to filter the Temurin release to retrieve.
```
Comment (olblak) [inline pkg/plugins/resources/temurin/spec.go:20]: Is it a required value?

File: pkg/plugins/resources/temurin/spec.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	// * "feature"
+	ReleaseLine string `yaml:",omitempty"`
+	// ReleaseType specifies the type of Temurin release to retrieve.
+	//
+	// default: "ga"
+	//
+	// Allowed values:
+	// * "ga" (General Availability)
+	// * "ea" (Early Availability, e.g. nightly builds)
+	ReleaseType string `yaml:",omitempty"`
```
Comment (olblak) [inline pkg/plugins/resources/temurin/spec.go:19]:
```suggestion
	ReleaseType string `yaml:",omitempty"  jsonschema:"required"`
```

If this field is required then this tag will trigger a warning if not set

Comment (dduportal) [reply pkg/plugins/resources/temurin/spec.go:19]: Oh thanks! Let me set up required/non required attributes!

**Review by olblak — APPROVED** [review]
Great pullrequest
Feel free to close my last comment if that field is not mandatory

File: pkg/plugins/resources/temurin/spec.go (hunk 4 above)
Comment (dduportal) [reply pkg/plugins/resources/temurin/spec.go:20]: Nope, it's not

File: pkg/plugins/resources/temurin/spec.go (hunk 5 above)
Comment (dduportal) [reply pkg/plugins/resources/temurin/spec.go:19]: Nevermind: field is not mandatory as we can have the following resource (with no spec at all):

```yaml
sources:
  ## Returns last LTS (implicit default)
  getLastDefaultVersion:
    kind: temurin
```

=> so no attribute mandatory

**Comment by dduportal** [comment]
> Great pullrequest Feel free to close my last comment if that field is not mandatory

Thanks! I let you merge if you feel like we can trigger a release

**Comment by dduportal** [comment]
PR doc: https://github.com/updatecli/website/pull/1545

**Pull request merged**